### PR TITLE
Add Component::printOutputInfo()

### PR DESCRIPTION
**Pull request by klshrinidhi**
Addressing part of #1503.

Sample output:
```
 Outputs from /World/internalSub
 ===============================
   subState
 
 Outputs from /World/Foo
 =======================
   BodyAcc
   Output1
   Output2
   Output3
   Qs
   return_by_ref
 
 Outputs from /World/Bar
 =======================
   PotentialEnergy
   activation
   copytesting
   copytestingMemVar
   fiberLength
 
 Outputs from /World/Foo2
 ========================
   BodyAcc
   Output1
   Output2
   Output3
   Qs
   return_by_ref
 
 Outputs from /World/InternalWorld/internalSub
 =============================================
   subState
 
 Outputs from /World/InternalWorld/Foo
 =====================================
   BodyAcc
   Output1
   Output2
   Output3
   Qs
   return_by_ref
 
 Outputs from /World/InternalWorld/Bar
 =====================================
   PotentialEnergy
   activation
   copytesting
   copytestingMemVar
   fiberLength
 
 Outputs from /World/InternalWorld/Foo2
 ======================================
   BodyAcc
   Output1
   Output2
   Output3
   Qs
   return_by_ref
 
 Outputs from /World/World3/internalSub
 ======================================
   subState
 
 Outputs from /World/World3/Foo
 ==============================
   BodyAcc
   Output1
   Output2
   Output3
   Qs
   return_by_ref
```

### CODE REVIEW

**Comment by chrisdembia** [comment]
I'll review this later, but what is `return_by_ref`?

**Comment by klshrinidhi** [comment]
> I'll review this later, but what is `return_by_ref`?

Not sure. The code is mostly copied from @tkuchida's code mentioned in #1503.

**Comment by tkuchida** [comment]
> The code is mostly copied from @tkuchida's code mentioned in #1503.

`return_by_ref` has nothing to do with me; it's an Output defined [here](https://github.com/opensim-org/opensim-core/blob/master/OpenSim/Common/Test/testComponentInterface.cpp#L150).

**Comment by klshrinidhi** [comment]
>  The code is mostly copied from @tkuchida's code mentioned in #1503.
>> `return_by_ref` has nothing to do with me; it's an Output defined [here](https://github.com/opensim-org/opensim-core/blob/master/OpenSim/Common/Test/testComponentInterface.cpp#L150).

Great. I didn't imply there was a problem.

File: OpenSim/Common/Component.cpp [hunk 1]
```diff
@@ -1524,6 +1504,31 @@ void Component::dumpConnections() const {
...
+
+        std::vector<std::string> outputNames = getOutputNames();
+        for (auto thisName : outputNames)
+            std::cout << "  " << thisName << std::endl;
+        std::cout << std::endl;
+    }
+
+    if (includeDescendants) {
+        for (const Component& thisComp : getComponentList<Component>()) {
+            // compList (comp's ComponentList) includes all descendants (i.e.,
```
Comment (tkuchida) [inline OpenSim/Common/Component.cpp]: There's no variable named `compList`. I'm guessing that you modified the code after C&P but didn't make the corresponding change to the comment.

Comment (klshrinidhi) [inline OpenSim/Common/Component.cpp]: Done.

File: OpenSim/Common/Component.h [hunk 2]
```diff
@@ -1488,6 +1488,10 @@ OpenSim_DECLARE_ABSTRACT_OBJECT(Component, Object);
         }
         std::cout << std::endl;
     }
+
+    /** Print outputs of this component and optionally, that of all
```
Comment (tkuchida) [inline OpenSim/Common/Component.h]: that -> those

**Comment by chrisdembia** [comment]
Cool stuff. 

I think it would be very helpful to add types: both of the Component, and for each of the outputs.

**Comment by klshrinidhi** [comment]
I have added type of the output to be printed as well.

```
 Outputs from /World/internalSub
 ===============================
   [double]  subState
 
 Outputs from /World/Foo
 =======================
   [SimTK::Vec<2,SimTK::Vec<3,double,1>,1>]  BodyAcc
                                   [double]  Output1
                   [SimTK::Vec<3,double,1>]  Output2
                                   [double]  Output3
                   [SimTK::Vector_<double>]  Qs
                                   [double]  return_by_ref
 
 Outputs from /World/Bar
 =======================
          [double]  PotentialEnergy
          [double]  activation
   [unsigned long]  copytesting
          [double]  copytestingMemVar
          [double]  fiberLength
 
 Outputs from /World/Foo2
 ========================
   [SimTK::Vec<2,SimTK::Vec<3,double,1>,1>]  BodyAcc
                                   [double]  Output1
                   [SimTK::Vec<3,double,1>]  Output2
                                   [double]  Output3
                   [SimTK::Vector_<double>]  Qs
                                   [double]  return_by_ref
 
 Outputs from /World/InternalWorld/internalSub
 =============================================
   [double]  subState
 
 Outputs from /World/InternalWorld/Foo
 =====================================
   [SimTK::Vec<2,SimTK::Vec<3,double,1>,1>]  BodyAcc
                                   [double]  Output1
                   [SimTK::Vec<3,double,1>]  Output2
                                   [double]  Output3
                   [SimTK::Vector_<double>]  Qs
                                   [double]  return_by_ref
 
 Outputs from /World/InternalWorld/Bar
 =====================================
          [double]  PotentialEnergy
          [double]  activation
   [unsigned long]  copytesting
          [double]  copytestingMemVar
          [double]  fiberLength
```

**Comment by klshrinidhi** [comment]
With the component type included:

```
 Outputs from /World/internalSub [Sub]
 =====================================
   [double]  subState
 
 Outputs from /World/Foo [Foo]
 =============================
   [SimTK::Vec<2,SimTK::Vec<3,double,1>,1>]  BodyAcc
                                   [double]  Output1
                   [SimTK::Vec<3,double,1>]  Output2
                                   [double]  Output3
                   [SimTK::Vector_<double>]  Qs
                                   [double]  return_by_ref
 
 Outputs from /World/Bar [Bar]
 =============================
          [double]  PotentialEnergy
          [double]  activation
   [unsigned long]  copytesting
          [double]  copytestingMemVar
          [double]  fiberLength
 
 Outputs from /World/Foo2 [Foo]
 ==============================
   [SimTK::Vec<2,SimTK::Vec<3,double,1>,1>]  BodyAcc
                                   [double]  Output1
                   [SimTK::Vec<3,double,1>]  Output2
                                   [double]  Output3
                   [SimTK::Vector_<double>]  Qs
                                   [double]  return_by_ref
 
 Outputs from /World/InternalWorld/internalSub [Sub]
 ===================================================
   [double]  subState
 
 Outputs from /World/InternalWorld/Foo [Foo]
 ===========================================
   [SimTK::Vec<2,SimTK::Vec<3,double,1>,1>]  BodyAcc
                                   [double]  Output1
                   [SimTK::Vec<3,double,1>]  Output2
                                   [double]  Output3
                   [SimTK::Vector_<double>]  Qs
                                   [double]  return_by_ref
 
 Outputs from /World/InternalWorld/Bar [Bar]
 ===========================================
          [double]  PotentialEnergy
          [double]  activation
   [unsigned long]  copytesting
          [double]  copytestingMemVar
          [double]  fiberLength
 
 Outputs from /World/InternalWorld/Foo2 [Foo]
 ============================================
   [SimTK::Vec<2,SimTK::Vec<3,double,1>,1>]  BodyAcc
                                   [double]  Output1
                   [SimTK::Vec<3,double,1>]  Output2
                                   [double]  Output3
                   [SimTK::Vector_<double>]  Qs
                                   [double]  return_by_ref
```

**Comment by tkuchida** [comment]
I suggest adding a (static const) map so that you can display something like
```
[SimTK::Vec<2,SimTK::Vec<3,double,1>,1> = Vec6]  BodyAcc
```
We would need to maintain this map over time, but in practice the number of relevant typedefs is probably manageable (and I think decoding "SimTK::Vec<2,SimTK::Vec<3,double,1>,1>", for example, would be helpful for users).

**Comment by chrisdembia** [comment]
To achieve what @tkuchida suggests, it would be neat if we could leverage an existing templated mechanism (
https://github.com/opensim-org/opensim-core/blob/master/OpenSim/Common/Object.h#L1083)
```
Object_GetClassName<type>::name()
```

But I don't think this would work, because it's compile-time.

**Comment by tkuchida** [comment]
We could try inferring the typedef by parsing "SimTK::Vec<2,SimTK::Vec<3,double,1>,1>" to grab the "2" and "3" and multiplying, but I'm not sure that will be true in general.

**Comment by klshrinidhi** [comment]
I have added type-alias to the printed output. 

```
 Outputs from /World/internalSub [Sub]
 =====================================
   [double]  subState
 
 Outputs from /World/Foo [Foo]
 =============================
   [SimTK::Vec<2,SimTK::Vec<3,double,1>,1> = SpatialVec]  BodyAcc
                                   [double]  Output1
                   [SimTK::Vec<3,double,1> = Vec3]  Output2
                                   [double]  Output3
                   [SimTK::Vector_<double> = Vector]  Qs
                                   [double]  return_by_ref
 
 Outputs from /World/Bar [Bar]
 =============================
          [double]  PotentialEnergy
          [double]  activation
   [unsigned long]  copytesting
          [double]  copytestingMemVar
          [double]  fiberLength
 
 Outputs from /World/Foo2 [Foo]
 ==============================
   [SimTK::Vec<2,SimTK::Vec<3,double,1>,1> = SpatialVec]  BodyAcc
                                   [double]  Output1
                   [SimTK::Vec<3,double,1> = Vec3]  Output2
                                   [double]  Output3
                   [SimTK::Vector_<double> = Vector]  Qs
                                   [double]  return_by_ref
 
 Outputs from /World/InternalWorld/internalSub [Sub]
 ===================================================
   [double]  subState
 
 Outputs from /World/InternalWorld/Foo [Foo]
 ===========================================
   [SimTK::Vec<2,SimTK::Vec<3,double,1>,1> = SpatialVec]  BodyAcc
                                   [double]  Output1
                   [SimTK::Vec<3,double,1> = Vec3]  Output2
                                   [double]  Output3
                   [SimTK::Vector_<double> = Vector]  Qs
                                   [double]  return_by_ref
 
 Outputs from /World/InternalWorld/Bar [Bar]
 ===========================================
          [double]  PotentialEnergy
          [double]  activation
   [unsigned long]  copytesting
          [double]  copytestingMemVar
          [double]  fiberLength
```

**Comment by tkuchida** [comment]
> I have added type-alias to the printed output.

Cool! I guess you also need to add the 84c7f04dd38913056e627326e236ea693a27b4b1 patch to the code that computes `maxlen`.

**Comment by klshrinidhi** [comment]
Sorry. I pushed too early. I have fixed the problem with number of spaces now.

```
 Outputs from /World/internalSub [Sub]
 =====================================
      [double]  subState
 
 Outputs from /World/Foo [Foo]
 =============================
   [SimTK::Vec<2,SimTK::Vec<3,double,1>,1> = SpatialVec]  BodyAcc
                                                [double]  Output1
                         [SimTK::Vec<3,double,1> = Vec3]  Output2
                                                [double]  Output3
                       [SimTK::Vector_<double> = Vector]  Qs
                                                [double]  return_by_ref
 
 Outputs from /World/Bar [Bar]
 =============================
             [double]  PotentialEnergy
             [double]  activation
      [unsigned long]  copytesting
             [double]  copytestingMemVar
             [double]  fiberLength
 
 Outputs from /World/Foo2 [Foo]
 ==============================
   [SimTK::Vec<2,SimTK::Vec<3,double,1>,1> = SpatialVec]  BodyAcc
                                                [double]  Output1
                         [SimTK::Vec<3,double,1> = Vec3]  Output2
                                                [double]  Output3
                       [SimTK::Vector_<double> = Vector]  Qs
                                                [double]  return_by_ref
 
 Outputs from /World/InternalWorld/internalSub [Sub]
 ===================================================
      [double]  subState
 
 Outputs from /World/InternalWorld/Foo [Foo]
 ===========================================
   [SimTK::Vec<2,SimTK::Vec<3,double,1>,1> = SpatialVec]  BodyAcc
                                                [double]  Output1
                         [SimTK::Vec<3,double,1> = Vec3]  Output2
                                                [double]  Output3
                       [SimTK::Vector_<double> = Vector]  Qs
                                                [double]  return_by_ref
 
 Outputs from /World/InternalWorld/Bar [Bar]
 ===========================================
             [double]  PotentialEnergy
             [double]  activation
      [unsigned long]  copytesting
             [double]  copytestingMemVar
             [double]  fiberLength
```

**Comment by chrisdembia** [comment]
Cool. LGTM. FYI The type names might be different with MSVC (the type names usually have `class ` prepended), and so the aliases might not get used there.

**Pull request merged**